compiler/src/iree/compiler/Utils/ADTExtras.h

```diff
@@ -0,0 +1,28 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_UTILS_ADTEXTRAS_H_
+#define IREE_COMPILER_UTILS_ADTEXTRAS_H_
+
+#include <utility>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/Sequence.h"
+
+namespace mlir::iree_compiler {
+
+template <typename FirstIteratee, typename... RestIteratees>
+auto enumerate_zip_equal(FirstIteratee&& first, RestIteratees&&... rest) {
+  size_t numElements =
+      std::distance(llvm::adl_begin(first), llvm::adl_end(first));
+  return llvm::zip_equal(llvm::seq(static_cast<size_t>(0), numElements),
+                         std::forward<FirstIteratee>(first),
+                         std::forward<RestIteratees>(rest)...);
```

Comment:
Are we going to upstream this to LLVM repo?

Reply:
Maybe eventually, but the bar would be higher and this is not the ideal way to do it. I think we would have to create a new `zippy` adaptor that maintains the index so that we don't have to check for the size ahead of time when assertions are disabled.
The practical difference that I can write this quick version that gets us 95% there in 5 minutes and a `zippy`-based implementation would require more thinking and arguing.